pkg/collector/corechecks/ebpf/tcp_queue_length.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	process_net "github.com/DataDog/datadog-agent/pkg/process/net"
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/tagger/collectors"
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	tcpQueueLengthCheckName = "tcp_queue_length"
+)
+
+// TCPQueueLengthConfig is the config of the TCP Queue Length check
+type TCPQueueLengthConfig struct {
+	CollectTCPQueueLength bool `yaml:"collect_tcp_queue_length"`
+	OnlyCountNbContexts   bool `yaml:"only_count_nb_contexts"` // For impact analysis only. To be removed after
+}
+
+// TCPQueueLengthCheck grabs TCP queue length metrics
+type TCPQueueLengthCheck struct {
+	core.CheckBase
+	instance *TCPQueueLengthConfig
+}
+
+func init() {
+	core.RegisterCheck(tcpQueueLengthCheckName, TCPQueueLengthFactory)
+}
+
+// TCPQueueLengthFactory is exported for integration testing
+func TCPQueueLengthFactory() check.Check {
+	return &TCPQueueLengthCheck{
+		CheckBase: core.NewCheckBase(tcpQueueLengthCheckName),
+		instance:  &TCPQueueLengthConfig{},
+	}
+}
+
+// Parse parses the check configuration and init the check
+func (t *TCPQueueLengthConfig) Parse(data []byte) error {
+	// default values
+	t.CollectTCPQueueLength = true
+	t.OnlyCountNbContexts = true
+
+	if err := yaml.Unmarshal(data, t); err != nil {
+		return err
+	}
+	return nil
+}
+
+//Configure parses the check configuration and init the check
+func (t *TCPQueueLengthCheck) Configure(config, initConfig integration.Data, source string) error {
+	// TODO: Remove that hard-code and put it somewhere else
+	process_net.SetSystemProbeSocketPath("/opt/datadog-agent/run/sysprobe.sock")
+
+	err := t.CommonConfigure(config, source)
+	if err != nil {
+		return err
+	}
+
+	return t.instance.Parse(config)
+}
+
+var tagsSet = make(map[string]struct{})
+
+// Run executes the check
+func (t *TCPQueueLengthCheck) Run() error {
+	if !t.instance.CollectTCPQueueLength {
+		return nil
+	}
+
+	sysProbeUtil, err := process_net.GetRemoteSystemProbeUtil()
+	if err != nil {
+		return err
+	}
+
+	data, err := sysProbeUtil.GetCheck("tcp_queue_length")
+	if err != nil {
+		return err
+	}
+
+	sender, err := aggregator.GetSender(t.ID())
+	if err != nil {
+		return err
+	}
+
+	for _, line := range data {
+		entityID := containers.BuildTaggerEntityName(line.ContainerID)
+		tags, err := tagger.Tag(entityID, collectors.OrchestratorCardinality)
+		if err != nil {
+			log.Errorf("Could not collect tags for container %s: %s", line.ContainerID, err)
+		}
+
+		tags = append(tags,
+			"saddr:"+line.Conn.Saddr.String(),
+			"daddr:"+line.Conn.Daddr.String(),
+			"sport:"+strconv.Itoa(int(line.Conn.Sport)),
+			"dport:"+strconv.Itoa(int(line.Conn.Dport)),
+			"pid:"+strconv.Itoa(int(line.Pid)))
```

Comment:
The cardinality of this will be super high. I don't know if you should loop in someone from metrics aggregation or just give a heads-up about this.

Reply:
That’s indeed my fear with this check.
That’s why there’s an option (it’s actually the default behaviour), to only count the number of contexts that this would generate.
Our plans was to gather figures before looping in the BE folks to see if it’s acceptable or not and then to loop in the customers to discuss what kind aggregation would make sense to them in order to reduce the number of context.

Reply:
That’s indeed my fear with this check.
That’s why there’s an option (it’s actually the default behaviour), to only count the number of contexts that this would generate.
Our plans was to gather figures before looping in the BE folks to see if it’s acceptable or not and then to loop in the customers to discuss what kind aggregation would make sense to them in order to reduce the number of context.